File: v4/rust/iprange-livedb/src/membership_delta.rs
//! Operation-private aggregation of dictionary refcount changes.

use crate::contract::{u32_le, PAGE_SIZE};
use crate::error::{Error, Result};
use crate::fixed_tree::{self, Codec, CursorDirection, CursorItem, RetiredPages, Store};
use crate::mapping::ByteSource;
use crate::slotted_page::Header;

const BRANCH_TYPE: u8 = 250;
const LEAF_TYPE: u8 = 251;
const AUX: u32 = 0x4d44_454c;
const ID_OFFSET: usize = 0;
const CHANGE_OFFSET: usize = 4;
const RECORD_SIZE: usize = 12;
const PENDING_SLOTS: usize = 2;

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub(crate) struct Delta {
    pub(crate) id: u32,
    pub(crate) change: i64,
}

#[derive(Clone, Copy, Debug, Default, PartialEq, Eq)]
pub(crate) struct Pending {
    slots: [Option<Delta>; PENDING_SLOTS],
}

impl Pending {
    pub(crate) const fn new() -> Self {
        Self {
            slots: [None; PENDING_SLOTS],
        }
    }

    pub(crate) fn is_empty(self) -> bool {
        self.slots.iter().all(Option::is_none)
    }
}

struct DeltaCodec;

pub(crate) struct Drain {
    cursor: fixed_tree::Cursor<DeltaCodec>,
}

struct DeltaItem;

impl Codec for DeltaCodec {
    type Key = u32;
    type Leaf = Delta;

    const BRANCH_TYPE: u8 = BRANCH_TYPE;
    const LEAF_TYPE: u8 = LEAF_TYPE;
    const AUX: u32 = AUX;
    const KEY_SIZE: usize = CHANGE_OFFSET;
    const LEAF_SIZE: usize = RECORD_SIZE;

    fn read_key<S: ByteSource>(cell: S, _level: u16) -> Result<Self::Key> {
        Ok(u32_le(cell, ID_OFFSET))
    }

    fn read_leaf<S: ByteSource>(cell: S) -> Result<Self::Leaf> {
        decode(cell)
    }

    fn write_key(key: Self::Key, output: &mut [u8]) {
        output[ID_OFFSET..CHANGE_OFFSET].copy_from_slice(&key.to_le_bytes());
    }
}

#[inline(always)]
pub(crate) fn track_buffered<S: Store>(
    store: &mut S,
    root: &mut u32,
    pending: &mut Pending,
    id: u32,
    change: i64,
) -> Result<()> {
    if id == 0 {
        return Ok(());
    }
    for slot in &mut pending.slots {
        if let Some(mut current) = *slot {
            if current.id != id {
                continue;
            }
            current.change = current
                .change
                .checked_add(change)
                .ok_or_else(|| Error::arithmetic_overflow("dictionary refcount delta"))?;
            *slot = Some(current);
            return Ok(());
        }
    }
    if let Some(slot) = pending.slots.iter_mut().find(|slot| slot.is_none()) {
        *slot = Some(Delta { id, change });
        return Ok(());
    }

    let oldest =
        pending.slots[0].ok_or_else(|| Error::corrupt("refcount delta pending slot is empty"))?;
    track(store, root, oldest.id, oldest.change)?;
    pending.slots[0] = pending.slots[1];
    pending.slots[1] = Some(Delta { id, change });
    Ok(())
}

pub(crate) fn flush<S: Store>(store: &mut S, root: &mut u32, pending: &mut Pending) -> Result<()> {
    for slot in &mut pending.slots {
        let Some(delta) = *slot else {
            continue;
        };
        track(store, root, delta.id, delta.change)?;
        *slot = None;
    }
    Ok(())
}

fn track<S: Store>(store: &mut S, root: &mut u32, id: u32, change: i64) -> Result<()> {
    crate::work::membership_delta_spill(1);
    let current = find(store, *root, id)?;
    if let Some(current) = current {
        if change == 0 {
            return Ok(());
        }
        let change = current
            .change
            .checked_add(change)
            .ok_or_else(|| Error::arithmetic_overflow("dictionary refcount delta"))?;
        let mut retired = RetiredPages::new();
        fixed_tree::mutate_leaf_u64::<DeltaCodec, S, _>(
            store,
            root,
            id,
            CHANGE_OFFSET,
            &mut retired,
            |_| Ok(fixed_tree::LeafU64Mutation::Replace(change as u64)),
        )?;
        require_private(retired)
    } else {
        insert(store, root, Delta { id, change })
    }
}

fn find<S: Store>(store: &S, root: u32, id: u32) -> Result<Option<Delta>> {
    if root == 0 {
        return Ok(None);
    }
    let found = fixed_tree::predecessor::<DeltaCodec, S>(store, root, id)?;
    match found {
        Some(found) if found.id == id => Ok(Some(found)),
        _ => Ok(None),
    }
}

fn insert<S: Store>(store: &mut S, root: &mut u32, delta: Delta) -> Result<()> {
    let mut retired = RetiredPages::new();
    if !fixed_tree::insert::<DeltaCodec, S>(store, root, &delta.encode(), &mut retired)? {
        return Err(Error::Corrupt("refcount delta key already exists"));
    }
    require_private(retired)
}

fn require_private(retired: RetiredPages) -> Result<()> {
    if retired.as_slice().is_empty() {
        Ok(())
    } else {
        Err(Error::Corrupt(
            "refcount delta tree contains a committed page",
        ))
    }
}

impl Drain {
    pub(crate) fn new<S: Store>(store: &mut S, root: u32) -> Result<Self> {
        Ok(Self {
            cursor: fixed_tree::Cursor::new_consuming(store, root, CursorDirection::Forward)?,
        })
    }

    pub(crate) fn next<S: Store>(&mut self, store: &mut S) -> Result<Option<Delta>> {
        self.cursor.next_consuming(store, &mut DeltaItem)
    }
}

impl CursorItem<DeltaCodec> for DeltaItem {
    type Output = Delta;

    fn read<S: ByteSource>(
        &mut self,
        page: S,
        header: &Header,
        _page_number: u32,
        index: usize,
    ) -> Result<Self::Output> {
        DeltaCodec::read_leaf(DeltaCodec::leaf_cell(page, header, index)?)
    }
}

fn decode<S: ByteSource>(cell: S) -> Result<Delta> {
    if cell.len() != RECORD_SIZE {
        return Err(Error::Corrupt("refcount delta record is malformed"));
    }
    let bytes = cell
        .array(CHANGE_OFFSET)
        .ok_or_else(|| Error::corrupt("refcount delta record is malformed"))?;
    Ok(Delta {
        id: u32_le(cell, ID_OFFSET),
        change: i64::from_le_bytes(bytes),
    })
}

impl Delta {
    fn encode(self) -> [u8; RECORD_SIZE] {
        let mut record = [0; RECORD_SIZE];
        record[ID_OFFSET..CHANGE_OFFSET].copy_from_slice(&self.id.to_le_bytes());
        record[CHANGE_OFFSET..].copy_from_slice(&self.change.to_le_bytes());
        record
    }
}

const _: () = assert!(DeltaCodec::LEAF_SIZE <= PAGE_SIZE);

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn signed_delta_encoding_round_trips() {
        for change in [i64::MIN, -7, -1, 0, 1, 7, i64::MAX] {
            let delta = Delta { id: 42, change };
            assert_eq!(decode(&delta.encode()).unwrap(), delta);
        }
    }
}

File: v4/go/internal/writer/membership_delta.go
// Operation-private aggregation of dictionary refcount changes (Rust
// membership_delta.rs): a two-slot pending buffer in front of one private
// fixed tree of 12-byte delta records. The draft edits accumulate
// refcount changes here; the workflow finish flushes the buffer and drains
// the tree in key order, retiring every private delta page through the
// consuming cursor.

package writer

import (
	"encoding/binary"

	"github.com/firehol/iprange/v4/go/internal/format"
	"github.com/firehol/iprange/v4/go/internal/tree"
	"github.com/firehol/iprange/v4/go/internal/work"
)

const (
	deltaBranchType   = format.PageType(250)
	deltaLeafType     = format.PageType(251)
	deltaAux          = uint32(0x4d44454c)
	deltaIDOffset     = 0
	deltaChangeOffset = 4
	deltaRecordSize   = 12
	deltaPendingSlots = 2
)

// memberDelta is one id refcount change (Rust Delta). The change is an
// i64 on the wire, sign included, exactly like the Rust record.
type memberDelta struct {
	id     uint32
	change int64
}

// writeMemberDelta encodes one delta record into dst (Rust Delta::encode).
// The caller owns the bounded target; the tree insert copies the bytes
// into the mapped page, so a caller-held scratch never allocates.
func writeMemberDelta(dst []byte, id uint32, change int64) {
	format.PutU32(dst[deltaIDOffset:deltaChangeOffset], id)
	binary.LittleEndian.PutUint64(dst[deltaChangeOffset:], uint64(change))
}

// deltaPending is the two-slot pending buffer of one draft (Rust
// Pending): slots hold [empty | one delta] by value; a full buffer
// spills the oldest slot into the delta tree. The value slots keep the
// batch merge allocation-free (Rust [Option<Delta>; PENDING_SLOTS]).
// scratch is the encode target of the spill inserts; it lives on the
// draft so tree inserts copy the record into the mapped page without a
// heap escape (the Go generic tree interface makes stack encodes
// escape).
type deltaPending struct {
	slots   [deltaPendingSlots]memberDelta
	used    [deltaPendingSlots]bool
	scratch [deltaRecordSize]byte
}

func newDeltaPending() deltaPending { return deltaPending{} }

// isEmpty reports no pending deltas (Rust Pending::is_empty).
func (p *deltaPending) isEmpty() bool {
	return !p.used[0] && !p.used[1]
}

// slot returns the used slot matching id, when present.
func (p *deltaPending) slot(id uint32) *memberDelta {
	for index := range p.used {
		if p.used[index] && p.slots[index].id == id {
			return &p.slots[index]
		}
	}
	return nil
}

// free returns the first empty slot, when present.
func (p *deltaPending) free() *memberDelta {
	for index := range p.used {
		if !p.used[index] {
			p.used[index] = true
			return &p.slots[index]
		}
	}
	return nil
}

// drop frees the slot at index.
func (p *deltaPending) drop(index int) {
	p.used[index] = false
	p.slots[index] = memberDelta{}
}

// deltaCodec is the wire contract of the delta tree (Rust DeltaCodec:
// fixed 4-byte id keys and 12-byte leaf records).
type deltaCodec struct{}

func (deltaCodec) BranchType() format.PageType { return deltaBranchType }
func (deltaCodec) LeafType() format.PageType   { return deltaLeafType }
func (deltaCodec) Aux() uint32                 { return deltaAux }
func (deltaCodec) KeySize() int                { return deltaChangeOffset }
func (deltaCodec) LeafSize() int               { return deltaRecordSize }

func (deltaCodec) ReadKey(cell []byte, _ uint16) (tree.Key, error) {
	if len(cell) < deltaChangeOffset {
		return tree.Key{}, corrupt("refcount delta key is truncated")
	}
	return tree.Key{Hi: uint64(format.U32(cell))}, nil
}

func (deltaCodec) ReadLeaf(cell []byte) (memberDelta, error) {
	return decodeMemberDelta(cell)
}

func (deltaCodec) WriteKey(key tree.Key, output []byte) {
	format.PutU32(output[deltaIDOffset:deltaChangeOffset], uint32(key.Hi))
}

func decodeMemberDelta(cell []byte) (memberDelta, error) {
	if len(cell) != deltaRecordSize {
		return memberDelta{}, corrupt("refcount delta record is malformed")
	}
	return memberDelta{
		id:     format.U32(cell[deltaIDOffset:deltaChangeOffset]),
		change: int64(binary.LittleEndian.Uint64(cell[deltaChangeOffset:])),
	}, nil
}

// trackBufferedDelta records one refcount change, merging equal ids in
// the pending slots and spilling the oldest slot when both are full
// (Rust track_buffered). id 0 (the empty bitmap) is never tracked.
func trackBufferedDelta(store tree.RetiringStore, root *uint32, pending *deltaPending, id uint32, change int64) error {
	if id == 0 {
		return nil
	}
	if slot := pending.slot(id); slot != nil {
		next, err := checkedDeltaChange(slot.change, change)
		if err != nil {
			return err
		}
		slot.change = next
		return nil
	}
	if slot := pending.free(); slot != nil {
		*slot = memberDelta{id: id, change: change}
		return nil
	}
	oldest := pending.slots[0]
	if !pending.used[0] {
		return corrupt("refcount delta pending slot is empty")
	}
	if err := trackMembershipDelta(store, root, oldest.id, oldest.change, pending.scratch[:]); err != nil {
		return err
	}
	pending.slots[0] = pending.slots[1]
	pending.used[0] = pending.used[1]
	pending.slots[1] = memberDelta{id: id, change: change}
	pending.used[1] = true
	return nil
}

// flushMembershipDeltas writes every pending slot into the delta tree
// (Rust flush).
func flushMembershipDeltas(store tree.RetiringStore, root *uint32, pending *deltaPending) error {
	for index := range pending.used {
		if !pending.used[index] {
			continue
		}
		slot := pending.slots[index]
		if err := trackMembershipDelta(store, root, slot.id, slot.change, pending.scratch[:]); err != nil {
			return err
		}
		pending.drop(index)
	}
	return nil
}

// trackMembershipDelta inserts or merges one delta record (Rust track: a
// merge is a u64 field edit at the change offset; a zero change on an
// existing record is a no-op).
func trackMembershipDelta(store tree.RetiringStore, root *uint32, id uint32, change int64, scratch []byte) error {
	work.MembershipDeltaSpill(1)
	current, found, err := findMemberDelta(store, *root, id)
	if err != nil {
		return err
	}
	if found {
		if change == 0 {
			return nil
		}
		next, err := checkedDeltaChange(current.change, change)
		if err != nil {
			return err
		}
		retired, _, err := tree.MutateLeafU64(deltaCodec{}, store, root, tree.Key{Hi: uint64(id)},
			deltaChangeOffset, tree.RetiredPages{}, func(leaf memberDelta) (tree.LeafU64Mutation, error) {
				return tree.LeafU64Mutation{DoReplace: true, Replace: uint64(next)}, nil
			})
		if err != nil {
			return err
		}
		return requirePrivateDeltaRetirement(&retired)
	}
	writeMemberDelta(scratch, id, change)
	retired, inserted, err := tree.Insert(deltaCodec{}, store, root, scratch[:deltaRecordSize], tree.RetiredPages{})
	if err != nil {
		return err
	}
	if !inserted {
		return corrupt("refcount delta key already exists")
	}
	return requirePrivateDeltaRetirement(&retired)
}

func checkedDeltaChange(current, change int64) (int64, error) {
	next := current + change
	if (change > 0 && next < current) || (change < 0 && next > current) {
		return 0, overflow("dictionary refcount delta")
	}
	return next, nil
}

func findMemberDelta(store tree.Store, root uint32, id uint32) (memberDelta, bool, error) {
	if root == 0 {
		return memberDelta{}, false, nil
	}
	found, ok, err := tree.Predecessor(deltaCodec{}, store, root, tree.Key{Hi: uint64(id)})
	if err != nil {
		return memberDelta{}, false, err
	}
	if !ok || found.id != id {
		return memberDelta{}, false, nil
	}
	return found, true, nil
}

func requirePrivateDeltaRetirement(retired *tree.RetiredPages) error {
	if retired.Len() == 0 {
		return nil
	}
	return corrupt("refcount delta tree contains a committed page")
}

// membershipDeltaDrain consumes one private delta tree in key order,
// retiring every page into the draft private stack through the consuming
// cursor (Rust Drain over Cursor::new_consuming).
type membershipDeltaDrain struct {
	cursor *tree.ForwardCursor[memberDelta]
}

func newMembershipDeltaDrain(store tree.RetiringStore, root uint32) (*membershipDeltaDrain, error) {
	cursor, err := tree.NewForwardCursor[memberDelta](deltaCodec{}, store, root, true)
	if err != nil {
		return nil, err
	}
	return &membershipDeltaDrain{cursor: cursor}, nil
}

// next returns the next delta in ascending id order (Rust
// next_consuming). The boolean is false at the end of the drain; this is
// a zero-allocation value return, unlike a pointer that would escape the
// cursor frame.
func (d *membershipDeltaDrain) next(store tree.RetiringStore) (memberDelta, bool, error) {
	delta, ok, err := d.cursor.Next()
	if err != nil {
		return memberDelta{}, false, err
	}
	return delta, ok, nil
}
